File: IOI/IOI 2009/Day 2/regions.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

#define HEAVY 625
#define MAXN 200000
#define MAXR 25000
#define MAXHEAVY 320

using namespace std;

struct Pair {
    int i, v;
    Pair(int i, int v) : i(i), v(v) {}
};

int N, R, Q, numHeavy = 0, numLight = 0;
int region[MAXN];
int regionCount[MAXR];
int heavyToAny[MAXHEAVY][MAXR];
int lightToHeavy[MAXR][MAXHEAVY];
int regionToNewI[MAXR];
vector<int> adj[MAXN];
vector<Pair> lightData[MAXR];

int heavyAnc[MAXHEAVY];
void precompHeavy(int n) {
    // Update this node's region count by # heavy ancestors
    for (int i = 0; i < numHeavy; i++) heavyToAny[i][region[n]] += heavyAnc[i];
    // Add this node (if appropriate) to ancestor count
    if (region[n] < numHeavy) heavyAnc[region[n]]++;
    for (auto e : adj[n]) {
        precompHeavy(e);
    }
    // Backtracking
    if (region[n] < numHeavy) heavyAnc[region[n]]--;
}

int numInST[MAXN][MAXHEAVY];
void precompLightToHeavy(int n) {
    // Find # nodes from each heavy region in subtree
    for (int e : adj[n]) {
        precompLightToHeavy(e);
        for (int i = 0; i < numHeavy; i++) numInST[n][i] += numInST[e][i];
    }
    // Update light to heavy query counts
    for (int i = 0; i < numHeavy; i++) {
        lightToHeavy[region[n]][i] += numInST[n][i];
    }
    // Add this node (if appropriate)
    if (region[n] < numHeavy) numInST[n][region[n]]++;
}

int clabel = 0;
void genLightData(int n) {
    lightData[region[n]].emplace_back(clabel++, 1);
    for (int e : adj[n]) genLightData(e);
    lightData[region[n]].emplace_back(clabel - 1, -1);
}

int calcLTL(int a, int b) {
    if (lightData[b].size() == 0) return 0;
    int currCount = 0, answer = 0;
    uint32_t bi = 0;
    for (auto const& p : lightData[a]) {
        while (lightData[b][bi].i <= p.i) {
            if (lightData[b][bi].v == 1) answer += currCount;
            bi++;
            if (bi == lightData[b].size()) return answer;
        }
        currCount += p.v;
    }
    return answer;
}

int main() {
    scanf("%d", &N);
    scanf("%d", &R);
    scanf("%d", &Q);
    int p, r;
    scanf("%d", &r);
    r--;
    region[0] = r;
    regionCount[r]++;
    for (int i = 1; i < N; i++) {
        scanf("%d", &p);
        scanf("%d", &r);
        p--;
        r--;
        region[i] = r;
        regionCount[r]++;
        adj[p].push_back(i);
    }

    // Figure out which regions are heavy
    for (int i = 0; i < R; i++) {
        if (regionCount[i] >= HEAVY) {
            // Heavy region
            regionToNewI[i] = numHeavy++;
        } else {
            // Light region
            regionToNewI[i] = R - 1 - numLight++;
        }
    }

    // Relabel regions in the new ordering (heavy regions are first)
    for (int i = 0; i < N; i++) region[i] = regionToNewI[region[i]];

    // Precomputation for (H, H) and (H, L) queries
    precompHeavy(0);

    // Precomputation for (L, H) [and technically (H, H)] queries
    precompLightToHeavy(0);

    // Precomputation for (L, L) queries
    for (int i = 0; i < R; i++) lightData[regionToNewI[i]].reserve(regionCount[i] * 2);
    genLightData(0);

    // Finally, answer the queries!
    int a, b;
    for (int i = 0; i < Q; i++) {
        scanf("%d", &a);
        scanf("%d", &b);
        a--;
        b--;
        a = regionToNewI[a];
        b = regionToNewI[b];
        if (a < numHeavy) {
            // Heavy to any
            printf("%d\n", heavyToAny[a][b]);
        } else if (b < numHeavy) {
            // Light to heavy
            printf("%d\n", lightToHeavy[a][b]);
        } else {
            // Light to light; calculate answer
            printf("%d\n", calcLTL(a, b));
        }
        fflush(stdout);
    }
}

File: IOI/IOI 2009/Day 2/regions.java
/*
Solution: Separate the regions into "light" and "heavy" regions. Then, do some precomp
to process queries involving at least 1 heavy region in O(1) time. Do light to light
queries online in O(N) time (where N is the # of nodes with those light regions).

Assuming a heavy region is >=500 nodes (motivated by subtasks).
Runtime: O(1000Q + 400N)
*/

import java.util.*;
import java.io.*;

@SuppressWarnings("unchecked")
public class regions {
    final int HEAVY = 500;
    int N, R, Q, numHeavy = 0, numLight = 0;
    int[] region, regionCount;
    int[][] heavyToAny, lightToHeavy;
    ArrayList<Integer>[] adj;
    ArrayList<Pair>[] lightData;

    regions(BufferedReader in, PrintWriter out) throws IOException {
        StringTokenizer st = new StringTokenizer(in.readLine());
        N = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());
        Q = Integer.parseInt(st.nextToken());
        adj = new ArrayList[N];
        for (int i = 0; i < N; i++) adj[i] = new ArrayList<>(2);

        region = new int[N];
        regionCount = new int[R];
        region[0] = Integer.parseInt(in.readLine()) - 1;
        regionCount[region[0]]++;
        int p;
        for (int i = 1; i < N; i++) {
            st = new StringTokenizer(in.readLine());
            p = Integer.parseInt(st.nextToken()) - 1;
            region[i] = Integer.parseInt(st.nextToken()) - 1;
            regionCount[region[i]]++;
            adj[p].add(i);
        }

        // Figure out which regions are heavy
        int[] regionToNewI = new int[R];
        for (int i = 0; i < R; i++) {
            if (regionCount[i] >= HEAVY) {
                // Heavy region
                regionToNewI[i] = numHeavy++;
            } else {
                // Light region
                regionToNewI[i] = R - 1 - numLight++;
            }
        }
        // System.out.println(Arrays.toString(regionToNewI));

        // Relabel regions in the new ordering (heavy regions are first)
        for (int i = 0; i < N; i++) region[i] = regionToNewI[region[i]];

        // Precomputation for (H, H) and (H, L) queries
        heavyToAny = new int[numHeavy][R];
        heavyAnc = new int[numHeavy];
        precompHeavy(0);

        // Precomputation for (L, H) [and technically (H, H)] queries
        lightToHeavy = new int[R][numHeavy];
        precompLightToHeavy(0);

        // Precomputation for (L, L) queries
        lightData = new ArrayList[R];
        for (int i = 0; i < R; i++) lightData[i] = new ArrayList<>(2);
        genLightData(0);

        // Finally, answer the queries!
        int a, b;
        for (int i = 0; i < Q; i++) {
            st = new StringTokenizer(in.readLine());
            a = regionToNewI[Integer.parseInt(st.nextToken()) - 1];
            b = regionToNewI[Integer.parseInt(st.nextToken()) - 1];
            if (a < numHeavy) {
                // Heavy to any
                // System.out.println("hta");
                out.println(heavyToAny[a][b]);
                out.flush();
            } else if (b < numHeavy) {
                // Light to heavy
                // System.out.println("lth");
                out.println(lightToHeavy[a][b]);
                out.flush();
            } else {
                // Light to light; calculate answer
                // System.out.println("ltl");
                out.println(calcLTL(a, b));
                out.flush();
            }
        }
    }

    int calcLTL(int a, int b) {
        ArrayList<Pair> aRanges = lightData[a];
        ArrayList<Pair> bRanges = lightData[b];
        if (bRanges.size() == 0) return 0;
        int currCount = 0, bi = 0, answer = 0;
        Pair p;
        for (Pair aRange : aRanges) {
            p = aRange;
            while (bRanges.get(bi).i <= p.i) {
                if (bRanges.get(bi).v == 1) answer += currCount;
                bi++;
                if (bi == bRanges.size()) return answer;
            }
            currCount += p.v;
        }
        return answer;
    }

    int[] heavyAnc;
    void precompHeavy(int n) {
        // Update this node's region count by # heavy ancestors
        for (int i = 0; i < numHeavy; i++) heavyToAny[i][region[n]] += heavyAnc[i];
        // Add this node (if appropriate) to ancestor count
        if (region[n] < numHeavy) heavyAnc[region[n]]++;
        for (int e : adj[n]) {
            precompHeavy(e);
        }
        // Backtracking
        if (region[n] < numHeavy) heavyAnc[region[n]]--;
    }

    int[] precompLightToHeavy(int n) {
        // Find # nodes from each heavy region in subtree
        int[] numInST = new int[numHeavy];
        for (int e : adj[n]) {
            int[] childST = precompLightToHeavy(e);
            for (int i = 0; i < numHeavy; i++) numInST[i] += childST[i];
        }
        // Update light to heavy query counts
        for (int i = 0; i < numHeavy; i++) {
            lightToHeavy[region[n]][i] += numInST[i];
        }
        // Add this node (if appropriate)
        if (region[n] < numHeavy) numInST[region[n]]++;
        return numInST;
    }

    int clabel = 0;
    void genLightData(int n) {
        lightData[region[n]].add(new Pair(clabel++, 1));
        for (int e : adj[n]) genLightData(e);
        lightData[region[n]].add(new Pair(clabel - 1, -1));
    }

    static class Pair {
        int i, v;
        Pair(int i, int v) {
            this.i = i;
            this.v = v;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);
        // BufferedReader in = new BufferedReader(new FileReader("regions.in"));
        // PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("regions.out")));
        new regions(in, out);
        in.close();
        out.close();
    }
}